### PR TITLE
FALCON-2199 Delete API support for extension job (user extension)

### PR DESCRIPTION
**Pull request by sandeepSamudrala**
This pull request is dependent on https://github.com/apache/falcon/pull/331

### CODE REVIEW

File: prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java [hunk 1]
```diff
@@ -273,29 +275,67 @@ public APIResult resume(@PathParam("job-name") String jobName,
...
+        return new APIResult(APIResult.Status.SUCCEEDED, "Extension job " + jobName + " deleted successfully");
+    }
+
+    private SortedMap<EntityType,List<Entity>> getJobEntities(String jobName, ExtensionJobsBean extensionJobsBean,
+                                                              String doAsUser) throws FalconException, IOException {
+        TreeMap<EntityType, List<Entity>> entityMap = new TreeMap<>();
+        ExtensionMetaStore metaStore = ExtensionStore.getMetaStore();
+
+        ExtensionType extensionType =  metaStore.getDetail(extensionJobsBean.getExtensionName()).getExtensionType();
+        if (ExtensionType.TRUSTED.equals(extensionType)) {
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: Why do this? The list of processes and feeds are stored in the Jobs table, even for trusted extensions.

File: prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java [hunk 2]
```diff
@@ -427,6 +467,28 @@ private BufferedRequest getBufferedRequest(HttpServletRequest request) {
         return new BufferedRequest(request);
     }
 
+
+    private void deleteEntities(SortedMap<EntityType, List<Entity>> entityMap, HttpServletRequest request)
+        throws IOException, JAXBException {
+        for (Map.Entry<EntityType, List<Entity>> entry : entityMap.entrySet()) {
+            for (final Entity entity : entry.getValue()) {
+                final HttpServletRequest bufferedRequest = getEntityStream(entity, entity.getEntityType(), request);
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: Move this also to the Util method as suggested in PR 331.

Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: Ack. Made the changes.

**Comment by sandeepSamudrala** [comment]
@pallavi-rao : Apologies . There were more changes into this which have been found during dev testing. I have made more changes. Please re-review it again. I have incorporated your review comments.

File: client/src/main/java/org/apache/falcon/ExtensionHandler.java [hunk 3]
```diff
@@ -185,8 +185,6 @@ private static String createStagePath(String extensionName, String jobName) {
                 for (File innerFile : files) {
                     if (innerFile.isFile()) {
                         urls.add(innerFile.toURI().toURL());
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/ExtensionHandler.java]: Don't want to support sub-directories under the build lib path? It is fine to take that stance, but, we will have to call that out explicitly in the documentation.

File: prism/src/main/java/org/apache/falcon/service/BacklogMetricEmitterService.java [hunk 4]
```diff
@@ -116,11 +116,14 @@ public void onRemove(Entity entity) throws FalconException{
         if (entity.getEntityType() != EntityType.PROCESS){
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/service/BacklogMetricEmitterService.java]: The check here could cause issues. Entity can be cluster or datasource. In which case, the explicit cast below will cause issues.

File: unit/src/test/java/org/apache/falcon/unit/TestFalconUnit.java [hunk 5]
```diff
@@ -436,34 +438,37 @@ public void testSubmitAndScheduleExtensionJob() throws Exception {
...
-        process = (Process)getClient().getDefinition(EntityType.PROCESS.toString(), "sample", null);
+        process = (Process) getClient().getDefinition(EntityType.PROCESS.toString(), "sample", null);
         Assert.assertEquals(process.getPipelines(), "testSample");
 
         apiResult = getClient().getStatus(EntityType.PROCESS, "sample", CLUSTER_NAME, null, false);
         assertStatus(apiResult);
         Assert.assertEquals(apiResult.getMessage(), "RUNNING");
+
+        apiResult = deleteExtensionJob(TEST_JOB, null);
+        assertStatus(apiResult);
```
Comment (pallavi-rao) [inline unit/src/test/java/org/apache/falcon/unit/TestFalconUnit.java]: Also, add asserts to ensure the Job nor its corresponding entities exist.

File: client/src/main/java/org/apache/falcon/ExtensionHandler.java (hunk 3 above)
Comment (sandeepSamudrala) [inline client/src/main/java/org/apache/falcon/ExtensionHandler.java]: Yes. There are occasional build failures too for recursive calling. I will add it to the documentation.

File: prism/src/main/java/org/apache/falcon/service/BacklogMetricEmitterService.java (hunk 4 above)
Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/service/BacklogMetricEmitterService.java]: Backlog is supported only for processes and so this check is there.

File: unit/src/test/java/org/apache/falcon/unit/TestFalconUnit.java (hunk 5 above)
Comment (sandeepSamudrala) [inline unit/src/test/java/org/apache/falcon/unit/TestFalconUnit.java]: Added the cases for the same.

**Comment by sandeepSamudrala** [comment]
Added Assert.fail post the api calls.

**Pull request closed without merging**